src/template.rs

```diff
@@ -0,0 +1,359 @@
+/// from a block in the `template!` macro.
+impl<B: Bundle> IntoIterator for Fragment<B> {
+    type Item = Box<dyn Prototype>;
+    type IntoIter = core::iter::Once<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        std::iter::once(Box::new(self) as Box<_>)
+    }
+}
+
+/// # Purpose
+///
+/// This macro gives you something a little like `jsx`. Much like jsx lets you build and compose html fragments
+/// at runtime using normal javascript functions, this lets you build and compose ECS hierarchy fragments
+/// in normal rust functions, using normal rust syntax.
+///
+/// Here's an example of what it looks like:
+/// ```rust
+/// # use i_cant_believe_its_not_bsn::*;
+/// # use bevy::prelude::*;
+/// # let dark_mode = false;
+/// # #[derive(Component)]
+/// # pub struct MyMarkerComponent;
+/// let template = template! {
+///     root: {(
+///         Text::new(""),
+///         TextFont::from_font_size(28.0),
+///         if dark_mode { TextColor::WHITE } else { TextColor::BLACK }
+///     )} [
+///         hello: { TextSpan::new("Hello") };
+///         world: { TextSpan::new("World") };
+///         punctuation: {( TextSpan::new("!"), MyMarkerComponent )};
+///     ];
+/// };
+/// ```
+///
+/// The grammer is simple: The template contains a list of nodes, each with a name. Each node
+/// may also have mote named nodes as children.
+///
+/// There is no custom syntax for logic. Every time you see `{ ... }` it's a normal rust code-block, and
+/// there are several places where you can substitute in code-blocked for fixed values.
+///
+/// The general format of a node is this:
+/// 1. The name (eg. `root:`) which may be a fixed symbol or a code-block, and which ends in a colon.
+/// 2. A code-block which returns a `Bundle` (eg. `{( TextSpan::new("!"), MyMarkerComponent )}`).
+/// 3. Optionally, a list of other nodes in square brackets.
+///
+/// You don't have to settle for a static structure either; instead of using the normal node syntax
+/// you can just plop in a codeblock which returns `IntoIterator<Item = Box<dyn Prototype>>`.
+///
+/// # Composition
+///
+/// It's easy to compose functions that return `Templates`.
+///
+/// ```rust
+/// # use i_cant_believe_its_not_bsn::*;
+/// # use bevy::prelude::*;
+/// # use bevy::color::palettes::css;
+/// fn hello_to(name: String, party_time: bool) -> Template {
+///     template! {
+///         greetings: { TextSpan::new("Hello") };
+///         name: {
+///             if party_time {
+///                 (TextSpan::new(name), TextColor::default())
+///             } else {
+///                 (TextSpan::new(format!("{}!!!!!", name)), TextColor(css::HOT_PINK.into()))
+///             }
+///         };
+///     }
+/// }
+/// ```
+///
+/// # Usage
+///
+/// Once you have a template, you can insert it into the world using `Commands::build`.
+///
+/// # Grammar
+///
+/// The entire `template!` macro is defined the the following ABNF grammar
```

Comment:
Ah, I edited this in earlier but it didn't quite make the commit
```suggestion
/// The entire `template!` macro is defined with the following ABNF grammar
```